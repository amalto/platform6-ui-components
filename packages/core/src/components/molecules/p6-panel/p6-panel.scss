$panel-border-color: $border !default;
$panel-border-width: $control-border-width !default;

:host {
  display: block;
  width: 100%;

  .panel-header {
    display: flex;
    flex-direction: row;
    align-items: center;
    border-color: $panel-border-color;
    border-width: $panel-border-width;
    border-bottom-style: solid;
    padding: $gap-size * 2 * 1px;

    ::slotted([slot='label']) {
      flex-grow: 3;
      font-family: $title-family;
    }
    ::slotted([slot='header-actions']) {
      display: flex;
      flex-direction: row;
      gap: $column-gap;
    }

    .panel-action--toggle {
      margin-left: $margin-gap * 2;
    }
  }

  .panel-body {
    visibility: visible;
    transition: height 0.42ms ease-in-out;
    padding: $gap-size * 2 * 1px;

    &.collapsed {
      height: 0;
      visibility: collapse;
    }
  }

  .panel-actions {
    border-color: $panel-border-color;
    border-width: $panel-border-width;
    border-top-style: solid;
  }
}
