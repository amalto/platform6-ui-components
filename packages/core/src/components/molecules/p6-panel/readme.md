# p6-panel

<!-- Auto Generated Below -->

## Properties

| Property   | Attribute  | Description            | Type      | Default |
| ---------- | ---------- | ---------------------- | --------- | ------- |
| `hideable` | `hideable` | Set the panel hideable | `boolean` | `false` |

## Dependencies

### Depends on

- [p6-action](../../atoms/p6-action)
- [p6-icon](../../atoms/p6-icon)

### Graph

```mermaid
graph TD;
  p6-panel --> p6-action
  p6-panel --> p6-icon
  style p6-panel fill:#f9f,stroke:#333,stroke-width:4px
```

---

_Built with [StencilJS](https://stenciljs.com/)_
