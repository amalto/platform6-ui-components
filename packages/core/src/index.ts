export { Components, JSX } from './components';
export { P6ButtonType } from './components/atoms/p6-button/p6-button';
export { P6CalendarType } from './components/atoms/p6-calendar/p6-calendar';
export { P6CheckboxValue } from './components/atoms/p6-checkbox/p6-checkbox';
export { P6InputValue } from './components/atoms/p6-input/p6-input';
export { P6InputType } from './components/atoms/p6-input/types';
export { Target } from './components/atoms/p6-link/p6-link';
export { P6RadioValue } from './components/atoms/p6-radio/p6-radio';
export { P6SelectControl, P6SelectValue } from './components/atoms/p6-select-native/p6-select-native';
export { P6SwitchValue } from './components/atoms/p6-switch/p6-switch';
export { P6TextareaValue } from './components/atoms/p6-textarea/p6-textarea';
export { P6LanguageValue } from './components/molecules/p6-language/p6-language';
export { P6TranslationControl, P6TranslationValue } from './components/molecules/p6-translation/p6-translation';
export {
  AlignColumnDetail,
  CellValueChangedDetail,
  Column,
  ColumnDefinition,
  ColumnId,
  DataItem,
  EditingCellDetail,
  FilterRowsDetail,
  MoveColumnDetail,
  ResetDefinitionsDetail,
  ResizeColumnDetail,
  Row,
  RowId,
  ShowColumnDetail,
  ShowOptionsDetail,
  SortColumnDetail,
} from './components/organisms/p6-grid/core/entities';
export { ServiceStatus } from './components/templates/p6-service-panel/service-status.enum';
export { P6Control } from './shared/form/control';
export { InvalidEventDetail, isCustomEvent, isInvalidEvent, isValidEvent, ValidEventDetail } from './shared/form/event';
export { Alignment, Mode, modes, Position, positions, Size, sizes, SortOrder } from './shared/types';
export { EnumArrayEntry, enumArrayToObject, enumToArray, EnumType } from './utils/enum';
export { LanguageCode } from './utils/language';
