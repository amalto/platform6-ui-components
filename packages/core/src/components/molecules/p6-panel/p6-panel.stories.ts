import { library } from '@fortawesome/fontawesome-svg-core';
import { faPlay, faStop } from '@fortawesome/free-solid-svg-icons';
import { ComponentProps, getElement, makeStory, Props } from '../../../shared/storybook';
import { Mode, Size } from '../../../shared/types';
import { P6Panel } from './p6-panel';

const component = 'p6-panel';

export default {
  title: 'Molecules/Panel',
  component,
};

library.add(faPlay, faStop);

const componentProps: ComponentProps = ['hideable', 'hidden'];

const getStoryField = (props?: Props<P6Panel>): HTMLElement => {
  return getElement(
    component,
    [
      getElement('span', 'Panel Header', {
        slot: 'label',
      }),
      getElement(
        'div',
        [
          getElement(
            'p6-button',
            [
              getElement('p6-icon', [], {
                name: 'play',
              }),
              'Start',
            ],
            { outlined: true, size: Size.small },
          ),
          getElement(
            'p6-button',
            [
              getElement('p6-icon', [], {
                name: 'stop',
              }),
              'Stop',
            ],
            { mode: Mode.danger, outlined: true, size: Size.small },
          ),
        ],
        {
          slot: 'header-actions',
        },
      ),
      getElement('p6-empty', []),
    ],
    props,
  );
};

export const Default = makeStory<{
  hideable: boolean;
}>({
  componentProps,
  args: {
    hideable: true,
  },
  builder: args => getStoryField(args),
});
