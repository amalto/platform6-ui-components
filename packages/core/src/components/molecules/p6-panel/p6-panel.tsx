import { Component, Element, h, Host, JSX, Prop, State } from '@stencil/core';
import { toArray } from '../../../utils/dom';

@Component({
  tag: 'p6-panel',
  styleUrl: 'p6-panel.scss',
  shadow: true,
})
export class P6Panel {
  @Element() host!: HTMLP6PanelElement;

  /**
   * Set the panel hideable
   */
  @Prop() hideable = false;

  @State() isHidden = false;

  render(): JSX.Element {
    const hasActions = toArray(this.host.children).filter(node => node.slot === 'actions').length > 0;

    return (
      <Host>
        <div class="panel-header">
          <slot name="label" />
          <slot name="header-actions" />
          {this.renderToggle(this.hideable, this.isHidden)}
        </div>
        <div class={{ 'panel-body': true, 'collapsed': this.isHidden }}>
          <slot />
        </div>
        {hasActions ? (
          <div class="panel-actions">
            <slot name="actions" />
          </div>
        ) : null}
      </Host>
    );
  }

  private renderToggle(hideable: boolean, hidden: boolean): JSX.Element {
    const icon = hidden ? 'angle-down' : 'angle-up';

    return hideable ? (
      <p6-action class="panel-action--toggle" onClick={this.onToggleHandler(!hidden)}>
        <p6-icon name={icon} />
      </p6-action>
    ) : null;
  }

  private onToggleHandler(hidden: boolean): (event: Event) => void {
    return () => {
      this.isHidden = hidden;
    };
  }
}
