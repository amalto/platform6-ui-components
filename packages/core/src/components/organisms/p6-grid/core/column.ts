import { Alignment, Direction, SortOrder } from '../../../../shared/types';
import { throwError } from '../../../../utils/error';
import { Column, ColumnDefinition, ColumnId, DataItem } from './entities';

export const MIN_WIDTH = 75;
export const DEFAULT_WIDTH = MIN_WIDTH;
export const INC_WIDTH = 10;

export function defaultGetValue<T extends DataItem>(data: T, context: ColumnDefinition<T>): string {
  if (context.field === undefined || data[context.field] === undefined) {
    return throwError(`field is undefined or doesn't exist, you should provide a 'getValue' function for ${JSON.stringify(context)}`);
  }

  return `${data[context.field as keyof T]}`;
}

export function defaultSetValue<T extends DataItem>(value: string, data: T, context: ColumnDefinition<T>): T {
  if (context.field === undefined || data[context.field] === undefined) {
    return throwError(`field is undefined or doesn't exist, you should provide a 'setValue' function for ${JSON.stringify(context)}`);
  }

  return {
    ...data,
    [context.field as keyof T]: value,
  };
}

let columnIds = 0;
export function fromDefinition<T extends DataItem>(column: ColumnDefinition<T>): Column<T> {
  // eslint-disable-next-line no-plusplus
  const id = column.id || column.field || columnIds++;

  return {
    ...column,
    id: `col-${id}`,
    editable: column.editable || false,
    hidden: column.hidden || false,
    width: column.width || 100,
    align: column.align || Alignment.center,
    sortOrder: column.sortOrder || SortOrder.none,
    disableHeaderMenu: column.disableHeaderMenu || false,
    getValue: column.getValue || defaultGetValue,
    setValue: column.setValue || defaultSetValue,
  };
}

export function isHidden<T extends DataItem>(column: Column<T>): boolean {
  return !!column.hidden;
}

export function replaceColumn<T extends DataItem>(updatedColumn: Column<T>, allColumns: Column<T>[]): Column<T>[] {
  return allColumns.map(col => (col.id === updatedColumn.id ? updatedColumn : col));
}

export function move<T extends DataItem>(id: ColumnId, direction: Direction, columns: Column<T>[]): Column<T>[] {
  const from = columns.findIndex(col => col.id === id);
  const to = from + (direction === Direction.left ? -1 : 1);

  if (from < 0 || to < 0 || to >= columns.length) {
    return columns;
  }

  const result = columns.slice();
  const item = result.splice(from, 1);

  return result.slice(0, to).concat(item).concat(result.slice(to, columns.length));
}
