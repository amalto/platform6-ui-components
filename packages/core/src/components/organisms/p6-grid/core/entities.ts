import { EventEmitter, JSX } from '@stencil/core';
import { Alignment, Direction, SortOrder } from '../../../../shared/types';

export type ColumnId = string;

export type DataItem = Record<string, unknown>;

export interface ColumnDefinition<T extends DataItem> {
  id?: ColumnId;
  label: string;
  field?: string | keyof T;
  editable?: boolean;
  filtreable?: boolean;
  sortable?: boolean;
  hidden?: boolean;
  width?: number;
  align?: Alignment;
  color?: string;
  sortOrder?: SortOrder;
  disableHeaderMenu?: boolean;
  getValue?: (data: T, context: ColumnDefinition<T>) => string;
  setValue?: (value: string, data: T, context: ColumnDefinition<T>) => T;
  cellEditor?: (data: T, context: ColumnDefinition<T>) => JSX.Element & CellEditor<T>;
  cellRenderer?: (data: T, rowId: RowId, context: ColumnDefinition<T>) => string | HTMLElement;
}

export interface Column<T extends DataItem> extends ColumnDefinition<T> {
  id: ColumnId;
  editable: boolean;
  hidden: boolean;
  width: number;
  align: Alignment;
  sortOrder: SortOrder;
  disableHeaderMenu: boolean;
  getValue: (data: T, context: ColumnDefinition<T>) => string;
  setValue: (value: string, data: T, context: ColumnDefinition<T>) => T;
}

export interface CellEditor<T extends DataItem> {
  column: Column<T>;
  row: Row<T>;
  value: string;

  p6GridCellValueChanged: EventEmitter<CellValueChangedDetail<T>>;
  p6GridEditingCell: EventEmitter<EditingCellDetail<T>>;
}
export type RowId = string;

export interface Row<T extends DataItem> {
  id: RowId;
  data: T;
}

interface ColumnEventDetail<T extends DataItem> {
  column: Column<T>;
}

export interface MoveColumnDetail<T extends DataItem> extends ColumnEventDetail<T> {
  direction: Direction;
}
export type ResizeColumnDetail<T extends DataItem> = ColumnEventDetail<T>;

export type AlignColumnDetail<T extends DataItem> = ColumnEventDetail<T>;

export type SortColumnDetail<T extends DataItem> = ColumnEventDetail<T>;

export type ShowColumnDetail<T extends DataItem> = ColumnEventDetail<T>;

export interface CellValueChangedDetail<T extends DataItem> {
  column: Column<T>;
  row: Row<T>;
  oldValue: string;
}

export enum EditingCellStatus {
  Start,
  Stop,
  Cancel,
}

export interface EditingCellDetail<T extends DataItem> {
  column: Column<T>;
  row: Row<T>;
  status: EditingCellStatus;
}

export interface ResetDefinitionsDetail {
  reset: boolean;
}

export interface ShowOptionsDetail {
  visible: boolean;
}

export interface FilterRowsDetail {
  value: string;
}
