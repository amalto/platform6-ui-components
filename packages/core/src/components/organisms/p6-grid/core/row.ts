import { SortOrder } from '../../../../shared/types';
import { Column, DataItem, Row, RowId } from './entities';

let rowIds = 0;
export function fromData<T extends DataItem>(data: T): Row<T> {
  rowIds += 1;

  return {
    id: `row-${rowIds}`,
    data,
  };
}

export function compareRow<T extends DataItem>(column: Column<T> | undefined): (a: Row<T>, b: Row<T>) => number {
  const sortOrder = column?.sortOrder || SortOrder.none;
  const factor = sortOrder === SortOrder.desc ? -1 : 1;

  return (a: Row<T>, b: Row<T>): number => {
    if (column === undefined || column.sortable === false || sortOrder === SortOrder.none) {
      return 0;
    }

    const valueA = column.getValue(a.data, column);
    const valueB = column.getValue(b.data, column);

    return valueA.localeCompare(valueB) * factor;
  };
}

export function filterBy<T extends DataItem>(searchValue: string, columns: Column<T>[]): (row: Row<T>) => boolean {
  return row => columns.filter(col => col.filtreable !== false).some(col => col.getValue(row.data, col).includes(searchValue));
}

export function rangeSelectRow<T extends DataItem>(currentRowIdSelect: RowId, selectedRows: Set<RowId>, displayedRow: Row<T>[]): Set<RowId> {
  const currentSelectedIndex = displayedRow.findIndex(row => row.id === currentRowIdSelect);

  if (currentSelectedIndex === -1) {
    return selectedRows;
  }

  const previousRowIdSelected = Array.from(selectedRows)
    .reverse()
    .find(rowId => displayedRow.findIndex(row => row.id === rowId) > -1);

  if (previousRowIdSelected === undefined) {
    return new Set(selectedRows.add(currentRowIdSelect));
  }

  const previousSelectedIndex = displayedRow.findIndex(row => row.id === previousRowIdSelected);

  let [from, to] = [previousSelectedIndex, currentSelectedIndex];
  let excludedFromSlice = currentRowIdSelect;
  if (currentSelectedIndex < previousSelectedIndex) {
    [from, to] = [to, from];
    excludedFromSlice = previousRowIdSelected;
  }

  return displayedRow
    .slice(from, to)
    .reduce((selected, cur) => {
      return cur.id === undefined ? selected : selected.add(cur.id);
    }, new Set(selectedRows))
    .add(excludedFromSlice);
}

export function replaceRow<T extends DataItem>(updatedRow: Row<T>, allRows: Row<T>[]): Row<T>[] {
  return allRows.map(row => (row.id === updatedRow.id ? updatedRow : row));
}
