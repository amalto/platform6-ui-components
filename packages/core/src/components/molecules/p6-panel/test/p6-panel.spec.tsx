import { newSpecPage } from '@stencil/core/testing';
import { P6Panel } from '../p6-panel';

describe('p6-panel', () => {
  it('renders', async () => {
    const page = await newSpecPage({
      components: [P6Panel],
      html: `<p6-panel></p6-panel>`,
    });
    expect(page.root).toEqualHtml(`
      <p6-panel>
        <mock:shadow-root>
          <slot></slot>
        </mock:shadow-root>
      </p6-panel>
    `);
  });
});
