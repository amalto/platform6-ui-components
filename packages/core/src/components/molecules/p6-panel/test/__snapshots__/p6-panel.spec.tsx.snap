// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`p6-panel renders 1`] = `
<p6-panel>
  <mock:shadow-root>
    <div class="panel-header">
      <slot name="label"></slot>
      <slot name="header-actions"></slot>
    </div>
    <div class="panel-body">
      <slot></slot>
    </div>
  </mock:shadow-root>
</p6-panel>
`;
